### ISSUE TITLE
Unable to eval clojurescript projects that depend on recent cljs features e.g. upstream deps and cljsjs/react

### ISSUE THREAD

**Issue by ThomasDeutsch** [LightTable/LightTable#1845]
When i run a script that requires a cljsjs dependency, i get this error:

```
clojure.lang.ExceptionInfo: No such namespace: cljsjs.react at line 1 file:/C:/Users/deutsch/.m2/repository/rum/rum/0.2.1/rum-0.2.1.jar!/rum.cljs :: {:tag :cljs/analysis-error, :file "file:/C:/Users/deutsch/.m2/repository/rum/rum/0.2.1/rum-0.2.1.jar!/rum.cljs", :line 1, :column 1}
                 core.clj:4403 clojure.core/ex-info
              analyzer.clj:299 cljs.analyzer/error
              analyzer.clj:296 cljs.analyzer/error
             analyzer.clj:1104 cljs.analyzer/analyze-deps
             analyzer.clj:1289 cljs.analyzer/eval3936[fn]
              MultiFn.java:249 clojure.lang.MultiFn.invoke
             analyzer.clj:1624 cljs.analyzer/analyze-seq
             analyzer.clj:1711 cljs.analyzer/analyze[fn]
             analyzer.clj:1704 cljs.analyzer/analyze
             analyzer.clj:1699 cljs.analyzer/analyze
             analyzer.clj:1880 cljs.analyzer/analyze-file[fn]
             analyzer.clj:1876 cljs.analyzer/analyze-file
             analyzer.clj:1101 cljs.analyzer/analyze-deps
             analyzer.clj:1289 cljs.analyzer/eval3936[fn]
              MultiFn.java:249 clojure.lang.MultiFn.invoke
             analyzer.clj:1624 cljs.analyzer/analyze-seq
             analyzer.clj:1711 cljs.analyzer/analyze[fn]
             analyzer.clj:1704 cljs.analyzer/analyze
             analyzer.clj:1699 cljs.analyzer/analyze
             analyzer.clj:1880 cljs.analyzer/analyze-file[fn]
             analyzer.clj:1876 cljs.analyzer/analyze-file
             analyzer.clj:1852 cljs.analyzer/analyze-file
```

lein cljsjsbuild 0.0.2740 will compile the code without errors, but i have to refresh the browser manually and this error message will pop up every time.

**Comment by narma** [comment]
It also happens on linux and seems it related to latest changes in clojurescript.

**Comment by narma** [comment]
As I see here https://groups.google.com/d/msg/clojurescript/rB5sj6vp3UU/qvlgHg_Yq0kJ needed to update clojurescript deps and repl code at lighttable for fix that issue.

**Comment by narma** [comment]
Also discussion from weasel tomjakubowski/weasel#37 with related problem

**Comment by aglie** [comment]
It also happens on Mac.

**Comment by narma** [comment]
Get LT repl works with this commit
https://github.com/narma/Clojure/commit/56afe890c0569e4a4e3015dddf72b2555028f97a
but it's not yet done solution, have tons of warnings in the console.

**Comment by aglie** [comment]
Thank you, I will try it the next days.

2015-02-04 10:52 GMT+01:00 Sergey Rublev <notifications@github.com>:

**Comment by aglie** [comment]
I applied the update to LightTable/plugins/clojure/lien-light-nrepl but I still get the same error. 

Is there a way to check which version of clojuresscript does the lein-light-nrepl use? Because when checking the *clojurescript-version* constant, I get the one which was used while compiling the project.

**Comment by 23min** [comment]
So is this a weasel issue? I logged a bug here too: https://github.com/adzerk/boot-cljs-repl/issues/7
Using clojurescript 2850

**Comment by narma** [comment]
@23min No, it's just a new API in clojurescript which are not supported by some repls, for example in Lighttable. Weasel maybe already works btw.

**Comment by 23min** [comment]
boot-cljs-repl uses Weasel under the hood, so I suspect it's the culprit.

**Comment by danjohansson** [comment]
Upgrading the clojurescript dep in the clojure plugin
and using this patch on cljs.clj worked for me.
```clojurescript
-(def compiler-env (cljs-env/default-compiler-env))
+(def compiler-env (let [ups-deps (cljs.closure/get-upstream-deps)
+                        opts {:ups-libs (:libs ups-deps)
+                              :ups-foreign-libs (:foreign-libs ups-deps)}]
+                    (cljs-env/default-compiler-env opts)))
```

**Comment by aglie** [comment]
Now with the new versions it magically worked for me. Thanks a lot!

**Comment by cldwalker** [comment]
Confirmed this is an issue when relying on cljsjs dependencies since our outdated clojurescript version is unable to pull in upstream deps. To reproduce, try evaling https://github.com/tonsky/rum/blob/gh-pages/examples/examples.cljs with an external browser.

@ibdknox Confirmed that https://github.com/narma/Clojure/commit/56afe890c0569e4a4e3015dddf72b2555028f97 fixes the upstream deps issue but using a new cljs version causes 100s of "Use of undeclared Var" warnings when compiling. Any pointers on what should be updated with [cljs.clj](https://github.com/LightTable/Clojure/blob/master/lein-light-nrepl/src/lighttable/nrepl/cljs.clj)?

**Comment by stephenbrady** [comment]
I saw the same above issues using Clojurescript 0.0-3126, which is the most recent now.  Applying the previously mentioned narma/Clojure@56afe89 cured the cljsjs error for me, but it exposed another issue.

When eval'ing cljs code via the LT nrepl, I was hitting the precondition assert in `with-core-cljs` in Clojurescript.  The fn signature to `with-core-cljs` changed back in September: https://github.com/clojure/clojurescript/commit/80ea15ed80f90669379dc08d40edd399fd415ae7

I made the following set of changes: https://github.com/fact/Clojure/compare/ccd9931c3e2eae87e139f85f9cfd99c278041d56...de8ef67d1314dd3800cf419e11fb17d17ccade05

Applying narma/Clojure@56afe89 and my own fixes, I was able to get a correctly working LT nrepl connection using the latest cljs (0.0-3126).

That said, I did not dig deeper into the change in `with-core-cljs` entails and which now takes an `opts` argument.  I just set `opts` to `nil`, which is already defined as the default value for the `opts` arg.

Hope that helps.

**Comment by jmatsushita** [comment]
Ran into this problem as well with chestnut and the atom-shell LightTable build. I've updated the ClojureScript versions in the Clojure plugin and LightTable and applied the 2 suggested fixes above but the compiler doesn't seem to be able to resolve dependencies and I get messages such as:
```
clojure.lang.ExceptionInfo: No such namespace: app.module, could not locate app/module.cljs at line 1 /.../src/cljs/app/core.cljs :: {:tag :cljs/analysis-error, :file "/.../src/cljs/app/app.cljs", :line 1, :column 1}
```

If I eval each dependency manually in the proper order then it works. Apologies if I'm missing something obvious as I'm quite new to this toolset.
